docs/functional/architecture/0004-Lifted_Stencils_with_Tuple_Return.md

```diff
@@ -0,0 +1,88 @@
+---
+tags: [iterator]
+---
+
+# Lifted Stencils with Tuple Return
+
+- **Status**: valid
+- **Authors**: Hannes Vogt (@havogt)
+- **Created**: 2022-02-16
+- **Updated**: 2022-02-16
+
+What's the return type of a lifted stencil that has tuple return type? Is it tuple of iterator or iterator of tuple?
+
+
+## Context
+
+Example
+
+```python
+def stencil():
+    return 42, 43
+
+# What is the type of
+lift(stencil)()
+```
+
+## Decision
+
+We return an iterator of tuples.
+
+There are several reasons why this is attractive:
+- At least for simple cases, the values returned from a stencil are most likely uniform and need to be treated uniformly, e.g. in a subsequent shift.
+- Implementation-wise, it is not straight-forward to implement tuple of iterators. E.g. on the Python-side the return type (tuple vs scalar) of the function is not known when `lift` is applied, i.e. when a lift-iterator needs to be constructed. The solution could be to execute-once arbitrary input then analyze the result (this would be ok, as our language doesn't have side-effects).
+- In general, transformations that remove unused elements are easier to implement than transformations that try to match things. See the following example.
+
+### Example of shifting all elements of the return
+
+Let's assume the following stencil
+
+```python
+def foo(inp0, inp1, inp2):
+    return deref(inp0), deref(inp1), deref(inp2)
+```
+
+### Iterator of tuples
+
+```python
+def uniform(inp0,inp1,inp2):
+    res = lift(foo)(inp0,inp1,inp2)
+    shifted = shift(O)(res)
+    return deref(shifted)
+
+def non_uniform(inp0,inp1,inp2):
+    res_it = lift(foo)(inp0,inp1,inp2)
+    shifted_vals_O = deref(shift(O)(res_it)) # everything is shifted
```

Comment:
Couldn't this also be written
```python
shifted_vals_0 = shift(0)(lift(lambda x: tuple_get(1, deref(x)))(res_it))
...
```

Which might not even require optimization and for which it is easy to provide a shorthand if needed?

Reply:
Makes sense. I'll add it as a note to the document.

Reply:
Actually, that wouldn't change anything, because lift is implemented as applying shift to the arguments of the function. It would be executed as

```python
shifted_vals_0 = (lambda x: tuple_get(1, deref(x)))(shift(0)(res_it)
```

Reply:
right.